modules/diskq/qdisk.c

```diff
@@ -569,6 +555,73 @@ qdisk_remove_head(QDisk *self)
   return TRUE;
 }
 
+static gboolean
+_overwrite_with_real_record_length(GString *serialized)
+{
+  guint32 record_length = GUINT32_TO_BE(serialized->len - sizeof(guint32));
+  if (record_length == 0)
+    return FALSE;
+
+  g_string_overwrite_len(serialized, 0, (gchar *) &record_length, sizeof(guint32));
+  return TRUE;
+}
+
+gboolean
+qdisk_serialize_msg(QDisk *self, LogMessage *msg, GString *serialized)
+{
+  gboolean result = TRUE;
+  SerializeArchive *sa = serialize_string_archive_new(serialized);
+
+  /* Leave space for the real record_length for later */
+  if (!serialize_write_uint32(sa, 0))
+    {
+      result = FALSE;
+      goto exit;
+    }
+
+  if (!log_msg_serialize(msg, sa, self->options->compaction ? LMSF_COMPACTION : 0))
+    {
+      result = FALSE;
+      goto exit;
+    }
+
+  if (!_overwrite_with_real_record_length(serialized))
+    {
+      msg_error("Error serializing empty message for the disk-queue file");
+      result = FALSE;
+      goto exit;
+    }
+
+exit:
+  if (!result)
+    g_string_truncate(serialized, 0);
+
+  serialize_archive_free(sa);
+
+  return result;
+}
+
+gboolean
+qdisk_deserialize_msg(QDisk *self, GString *serialized, LogMessage **msg)
+{
+  gboolean result = TRUE;
+  SerializeArchive *sa = serialize_string_archive_new(serialized);
+  *msg = log_msg_new_empty();
+
+  if (!log_msg_deserialize(*msg, sa))
+    {
+      msg_error("Can't read correct message from disk-queue file",
+                evt_tag_str("filename", qdisk_get_filename(self)),
+                evt_tag_long("read_position", qdisk_get_reader_head(self)));
```

Comment:
read_position might not be the best to be used here.

Reply:
Removed.

Reply:
What is the issue using `read_position` here ?

Reply:
It is not guaranteed, that the `deserialize` function will be called on a message, that was read from the `read_head`.

Reply:
what if you move the error message from here, where you can be sure about that ? as it seems like useful information to have, and error reporting separation from logic itself not so bad idea otherwise as well.

Reply:
I added another error message to the caller site. I think having 2 errors is not a big problem, this problem should rarely occur.